pkg/services/ngalert/state/manager.go

```diff
@@ -314,41 +312,79 @@ func (st *Manager) flushState(ctx context.Context) {
 	st.cache.mtxStates.Lock()
 	defer st.cache.mtxStates.Unlock()
 	totalStates, errorsCnt := 0, 0
+	var stateSlice []*State
 	for _, orgStates := range st.cache.states {
 		for _, ruleStates := range orgStates {
+			stateSlice = stateSlice[:0]
 			for _, state := range ruleStates {
-				err := st.saveState(ctx, state)
-				totalStates++
-				if err != nil {
-					st.log.Error("failed to save alert state", append(state.GetRuleKey().LogContext(), "labels", state.Labels.String(), "state", state.State.String(), "err", err.Error()))
-					errorsCnt++
-				}
+				stateSlice = append(stateSlice, state)
+			}
+			savedCount, failedCount := st.saveAlertStates(ctx, stateSlice...)
+			if failedCount != 0 {
+				st.log.Error("failed to save alert states.", "count", failedCount)
```

Comment:
I think we log this twice, here and https://github.com/grafana/grafana/pull/54072/files#diff-6420fa5148996efbf754b18838902601c17b758ab7da6279dc019b5b143fd5c2R375